modules/apps/portal/portal-http-impl/src/main/java/com/liferay/portal/http/internal/HttpImpl.java

```diff
@@ -451,6 +452,18 @@ protected void modified(Map<String, Object> properties) {
 				HttpConfiguration.class, properties);
 
 		_keepAliveTimeout = httpConfiguration.keepAliveTimeout();
+
+		PoolingHttpClientConnectionManager poolingHttpClientConnectionManager =
+			_poolingHttpClientConnectionManagerDCLSingleton.getSingleton(
+				HttpImpl::_createPoolingHttpClientConnectionManager);
+
+		if (httpConfiguration.tcpKeepAliveEnabled()) {
+			poolingHttpClientConnectionManager.setDefaultSocketConfig(
+				_keepAliveSocketConfig);
+		}
+		else {
+			poolingHttpClientConnectionManager.setDefaultSocketConfig(null);
+		}
 	}
 
```

Comment:
@ericyanLr, the else bloc is unnecessary because its default value is null.

In addition, I think we place the [logic](https://github.com/liferay-core-infra/liferay-portal/pull/6371/commits/c211861727950f8a2268a6b46691846acd4c95df) to `modified()` which can be more clear. Only when `_KEEPALIVE_TIMEOUT > 0`, we instantiate `new ConnectionKeepAliveStrategy`.

Reply:
Hi @yuhai ,

From my understanding, both of these mentioned logic blocks are necessary to handle scenarios where the configurations are changed in System Settings via runtime.

**Regarding:**
> the else block is unnecessary because its default value is null.

An example of a scenario would be:
1. Startup Liferay
2. Navigate to Control Panel > System Settings >> Infrastructure >> HTTP Configuration
    - Check `tcpKeepAliveEnabled`
    - Save
3. Navigate to Control Panel > Server Administration >> Script
    - Run the following script to utilize the HTTP component:
      ```
      println com.liferay.portal.kernel.util.HttpUtil.URLtoString("http://localhost:8080");
      ```
    - Verify SO_KEEPALIVE is being utilized (TCP Keep Alive is being used)
      - In Linux, we can check this by running the command: `netstat -p tcp -tano`
4. Navigate to Control Panel > System Settings >> Infrastructure >> HTTP Configuration
    - Uncheck `tcpKeepAliveEnabled`
    - Save
5. Repeat Step 3
6. **Expected Result:** TCP Keep Alive is no longer being used for requests

From my understanding, if we do not include the else block, we would continue to utilize our `_keepAliveSocketConfig`, which enables TCP Keep Alive, even though we unchecked the `tcpKeepAliveEnabled` setting.
- When modifying the OSGI configuration, the objects in the component are left unchanged, meaning the `poolingHttpClientConnectionManager` would continue to use our `_keepAliveSocketConfig`, unless we change it when the modified method is called.

**Regarding:**
> In addition, I think we place the [logic](https://github.com/liferay-core-infra/liferay-portal/pull/6371/commits/c211861727950f8a2268a6b46691846acd4c95df) to `modified()` which can be more clear. Only when `_KEEPALIVE_TIMEOUT > 0`, we instantiate `new ConnectionKeepAliveStrategy`.

An example of a scenario would be:
1. Startup Liferay
3. Navigate to Control Panel > Server Administration >> Script
    - Run the following script to utilize the HTTP component:
      ```
      println com.liferay.portal.kernel.util.HttpUtil.URLtoString("http://localhost:8080");
      ```
4. Navigate to Control Panel > System Settings >> Infrastructure >> HTTP Configuration
    - Set `keepAliveTimeout` to `10`
    - Save
5. Repeat Step 2
6. **Expected Result:** Our `keepAliveTimeout` should be taken into consideration and the logic in our custom `ConnectionKeepAliveStrategy` is being executed

From my understanding, if we do not instantiate our custom `ConnectionKeepAliveStrategy` from the beginning, it may end up never being utilized, due to our usage of DCLSingleton.
- Since we rely on the method: `_createCloseableHttpClient()` and `_createProxyCloseableHttpClient`, these are only called by: `getCloseableHttpClient`
  ``` 
  protected CloseableHttpClient getCloseableHttpClient(HttpHost proxyHost) {
	if (proxyHost != null) {
		return _proxyCloseableHttpClientDCLSingleton.getSingleton(
			this::_createProxyCloseableHttpClient);
	}

	return _closeableHttpClientDCLSingleton.getSingleton(
		this::_createCloseableHttpClient);
  }
  ```
- And since it's a singleton, this logic would not be called again until the singleton is destroyed, and so our new instantiated ConnectionKeepAliveStrategy would not be utilized until that happens. It would instead continue to utilize the default, which is null.

---

Here is also a link to the branch I used to test your suggested changes:
- Branch: https://github.com/ericyanLr/liferay-portal/tree/LPS-203867-10_requested-changes
- Compare: https://github.com/ericyanlr/liferay-portal/compare/master...ericyanLr:liferay-portal:LPS-203867-10_requested-changes

Can you help take a look at my response above and let me know if there's anything I may have missed or misunderstood?

Thanks!

Reply:
Hi @yuhai ,

All good, thanks for helping to take a look!

Regarding your suggestion, one scenario that would be affected, would be if there is an HTTP Request in progress (e.g. downloading a file, or waiting for a response from a slow website) and the deactivate method is called; those requests would result in an error: `java.lang.IllegalStateException: Connection pool shut down`.
- This is because the `deactivate` method would end up calling: `poolingHttpClientConnectionManager.shutdown()`

**Question**
If we destroy the singletons, I think it would be difficult to maintain the existing requests. Can you help take a look at this scenario and let me know if you would still like to proceed with this path?

Thanks!

Reply:
Hi @ericyanLr,

I thought `poolingHttpClientConnectionManager.shutdown()` is expected when modifying service. As I can see, after moving portal.properties to osgi configuration, we also need to set `.max.connections.per.host` and `.max.total.connections` to `PoolingHttpClientConnectionManager`. For existing request, if we only change same object `PoolingHttpClientConnectionManager` status/properties, I am not sure if they are affected. But as you mentioned issue, it is caused by our current change. I think you may firstly use your previous way that using same  `PoolingHttpClientConnectionManager`, instead of using `_poolingHttpClientConnectionManagerDCLSingleton.destroy()`.

For `_proxyCloseableHttpClientDCLSingleton `and `_closeableHttpClientDCLSingleton`, I think we need to invoke their `destroy()` when modifying service because the property `.timeout` need to be reset.

We still keep the logic that instantiate our `own ConnectionKeepAliveStrategy()` when keepAliveTimeout > 0.

Please let me know if you have further concerns.

Thanks.
Hai

Reply:
Hi @ericyanLr,

I just have a concern.

> For _proxyCloseableHttpClientDCLSingleton and _closeableHttpClientDCLSingleton, I think we still need to invoke their destroy() when modifying service because the property .timeout need to be reset. And they seem can't follow with PoolingHttpClientConnectionManager way.

The timeout property is set `RequestConfig`, and then using `proxyHttpClientBuilder.build()` set it to `InternalHttpClient#defaultConfig`. I don't see the timeout property is related with `PoolingHttpClientConnectionManager`.

We may further discuss based on your new pr if necessary.

Thanks,
Hai

Reply:
Hi @yuhai ,

I took a closer look today and now understand what you mean. I agree, we will need to destroy the `_proxyCloseableHttpClientDCLSingleton` and `_closeableHttpClientDCLSingleton` when the service is modified.

**Game Plan**
I'm working on finishing up the migrations + adding tests and will also include this change as well. Afterwards, I'll send a follow-up PR.

As discussed, the new PR will incorporate the following new changes:
- Do not need to call the deactivate method when the service is modified
  - Meaning, we will not be destroying / shutting down any existing PoolingHttpClientConnectionManager
- Destroy the `_proxyCloseableHttpClientDCLSingleton` and `_closeableHttpClientDCLSingleton` when the service is modified
- Migrate the relevant portal properties to OSGI configurations
- Include the discussed optimizations
  - e.g. We still keep the logic that instantiate our own ConnectionKeepAliveStrategy() when keepAliveTimeout > 0.

I'll close out this PR and resend a new one when ready.

Feel free to let me know if you have any further questions.
Thanks again for all your help!